Source/Core/Common/TraversalClient.cpp

```diff
@@ -0,0 +1,360 @@
+{
+	if (from->host == m_ServerAddress.host &&
+	    from->port == m_ServerAddress.port)
+	{
+		if (size < sizeof(TraversalPacket))
+		{
+			ERROR_LOG(NETPLAY, "Received too-short traversal packet.");
+		}
+		else
+		{
+			HandleServerPacket((TraversalPacket*) data);
+			return true;
+		}
+	}
+	return false;
+}
+
+//--Temporary until more of the old netplay branch is moved over
+void TraversalClient::Update()
+{
+	ENetEvent netEvent;
+	if (enet_host_service(m_NetHost, &netEvent, 4) > 0)
+	{
+		switch (netEvent.type)
+		{
+		case ENET_EVENT_TYPE_RECEIVE:
+			TestPacket(netEvent.packet->data, netEvent.packet->dataLength, &netEvent.peer->address);
+
+			enet_packet_destroy(netEvent.packet);
+			break;
+		default:
+			break;
+		}
+	}
+	HandleResends();
+}
+
+void TraversalClient::HandleServerPacket(TraversalPacket* packet)
+{
+	u8 ok = 1;
+	switch (packet->type)
+	{
+	case TraversalPacketAck:
+		if (!packet->ack.ok)
+		{
+			OnFailure(ServerForgotAboutUs);
+			break;
+		}
+		for (auto it = m_OutgoingTraversalPackets.begin(); it != m_OutgoingTraversalPackets.end(); ++it)
+		{
+			if (it->packet.requestId == packet->requestId)
+			{
+				m_OutgoingTraversalPackets.erase(it);
+				break;
+			}
+		}
+		break;
+	case TraversalPacketHelloFromServer:
+		if (m_State != Connecting)
+			break;
+		if (!packet->helloFromServer.ok)
+		{
+			OnFailure(VersionTooOld);
+			break;
+		}
+		m_HostId = packet->helloFromServer.yourHostId;
+		m_State = Connected;
+		if (m_Client)
+			m_Client->OnTraversalStateChanged();
+		break;
+	case TraversalPacketPleaseSendPacket:
+		{
+		// security is overrated.
+		ENetAddress addr = MakeENetAddress(&packet->pleaseSendPacket.address);
+		if (addr.port != 0)
+		{
+			char message[] = "Hello from Dolphin Netplay...";
+			ENetBuffer buf;
+			buf.data = message;
+			buf.dataLength = sizeof(message) - 1;
+			enet_socket_send(m_NetHost->socket, &addr, &buf, 1);
+		}
+		else
+		{
+			// invalid IPV6
+			ok = 0;
+		}
+		break;
+		}
+	case TraversalPacketConnectReady:
+	case TraversalPacketConnectFailed:
+		{
+		if (!m_PendingConnect || packet->connectReady.requestId != m_ConnectRequestId)
+			break;
+
+		m_PendingConnect = false;
+
+		if (!m_Client)
+			break;
+
+		if (packet->type == TraversalPacketConnectReady)
+			m_Client->OnConnectReady(MakeENetAddress(&packet->connectReady.address));
+		else
+			m_Client->OnConnectFailed(packet->connectFailed.reason);
+		break;
+		}
+	default:
+		WARN_LOG(NETPLAY, "Received unknown packet with type %d", packet->type);
+		break;
+	}
+	if (packet->type != TraversalPacketAck)
+	{
+		TraversalPacket ack = {};
+		ack.type = TraversalPacketAck;
+		ack.requestId = packet->requestId;
+		ack.ack.ok = ok;
+
+		ENetBuffer buf;
+		buf.data = &ack;
+		buf.dataLength = sizeof(ack);
+		if (enet_socket_send(m_NetHost->socket, &m_ServerAddress, &buf, 1) == -1)
+			OnFailure(SocketSendError);
+	}
+}
+
+void TraversalClient::OnFailure(int reason)
```

Comment:
`int reason` might want to be changed to `FailureReason reason`. This way you can also eliminate the default case in the switch as well.